From b4e32fa5c98e97e02fa0330fa5526f4170ee0f4c Mon Sep 17 00:00:00 2001
From: Bing Han <bing.han@intel.com>
Date: Mon, 14 Jun 2021 13:07:59 +0530
Subject: [PATCH] btusb: add prepare suspend blocker

Prepare suspend blocker is added to prevent system going to suspend
when BT firmware download is ongoing

Tracked-On: OAM-97127
Signed-off-by: Bing Han <bing.han@intel.com>
---
 drivers/bluetooth/btusb.c | 21 +++++++++++++++++----
 drivers/usb/core/usb.c    | 28 +++++++++++++++++++++++++++-
 include/linux/usb.h       |  1 +
 3 files changed, 45 insertions(+), 5 deletions(-)

diff --git a/drivers/bluetooth/btusb.c b/drivers/bluetooth/btusb.c
index 974b5ec43485..5ff2bde817d5 100644
--- a/drivers/bluetooth/btusb.c
+++ b/drivers/bluetooth/btusb.c
@@ -492,6 +492,8 @@ static const struct dmi_system_id btusb_needs_reset_resume_table[] = {
 #define BTUSB_TX_WAIT_VND_EVT	13
 #define BTUSB_WAKEUP_DISABLE	14
 #define BTUSB_USE_ALT1_FOR_WBS	15
+#define BTUSB_BLOCK_SUSPEND    16
+
 
 /* Per core spec 5, vol 4, part B, table 2.1,
  * list the hci packet payload sizes for various ALT settings.
@@ -2484,7 +2486,7 @@ static int btusb_intel_download_firmware_newgen(struct hci_dev *hdev,
                         /* Firmware has already been loaded */
                         set_bit(BTUSB_FIRMWARE_LOADED, &data->flags);
                         return 0;
-                }
+		}
 		bt_dev_err(hdev, "Failed to load Intel firmware file %s (%d)",
 			   fwname, err);
 		return err;
@@ -2809,11 +2811,9 @@ static int btusb_setup_intel_new(struct hci_dev *hdev)
 	/* controller is already having an operational firmware */
 	if (ver.fw_variant == 0x23)
 		goto finish;
-
 	err = btusb_intel_boot(hdev, boot_param);
 	if (err)
 		return err;
-
 	clear_bit(BTUSB_BOOTLOADER, &data->flags);
 
 	err = btusb_setup_intel_new_get_fw_name(&ver, &params, ddcname,
@@ -2867,7 +2867,7 @@ static int btusb_setup_intel_new(struct hci_dev *hdev)
 	 * and thus no need to fail the setup.
 	 */
 	btintel_set_event_mask(hdev, false);
-
+	clear_bit(BTUSB_BLOCK_SUSPEND, &data->flags);
 	return 0;
 }
 
@@ -4428,6 +4428,8 @@ static int btusb_probe(struct usb_interface *intf,
 		data->recv_event = btusb_recv_event_intel;
 		data->recv_bulk = btusb_recv_bulk_intel;
 		set_bit(BTUSB_BOOTLOADER, &data->flags);
+	        set_bit(BTUSB_BLOCK_SUSPEND, &data->flags);
+
 	} else {
 		data->recv_event = btusb_recv_evt;
 		data->recv_bulk = btusb_recv_bulk;
@@ -4765,6 +4767,16 @@ static void btusb_disconnect(struct usb_interface *intf)
 }
 
 #ifdef CONFIG_PM
+static int btusb_prepare(struct usb_interface *intf)
+{
+	struct btusb_data *data = usb_get_intfdata(intf);
+	if (test_bit(BTUSB_BLOCK_SUSPEND, &data->flags)) {
+		return -EBUSY;
+	}
+	return 0;
+}
+
+
 static int btusb_suspend(struct usb_interface *intf, pm_message_t message)
 {
 	struct btusb_data *data = usb_get_intfdata(intf);
@@ -4912,6 +4924,7 @@ static struct usb_driver btusb_driver = {
 	.probe		= btusb_probe,
 	.disconnect	= btusb_disconnect,
 #ifdef CONFIG_PM
+	.prepare        = btusb_prepare,
 	.suspend	= btusb_suspend,
 	.resume		= btusb_resume,
 #endif
diff --git a/drivers/usb/core/usb.c b/drivers/usb/core/usb.c
index f16c26dc079d..4cee848d2b23 100644
--- a/drivers/usb/core/usb.c
+++ b/drivers/usb/core/usb.c
@@ -451,7 +451,33 @@ static int usb_dev_uevent(struct device *dev, struct kobj_uevent_env *env)
 
 static int usb_dev_prepare(struct device *dev)
 {
-	return 0;		/* Implement eventually? */
+	struct usb_device *udev = to_usb_device(dev);
+	struct usb_interface *intf;
+	struct usb_driver *driver;
+	int status = 0;
+	int i = 0, n = 0;
+
+	if (udev->state == USB_STATE_NOTATTACHED)
+		return 0;
+
+	if (udev->actconfig) {
+		n = udev->actconfig->desc.bNumInterfaces;
+		dev_info(dev, "bNumInterfaces %d\n", n);
+		for (i = n - 1; i >= 0; --i) {
+			intf = udev->actconfig->interface[i];
+			if (intf->condition == USB_INTERFACE_UNBOUND)
+				continue;
+			driver = to_usb_driver(intf->dev.driver);
+			if (driver->prepare) {
+				dev_info(dev, "driver->prepare\n");
+				status = driver->prepare(intf);
+				dev_info(dev, "driver->prepare status: %d\n", status);
+				if (status)
+					return status;
+			}
+		}
+	}
+	return 0;
 }
 
 static void usb_dev_complete(struct device *dev)
diff --git a/include/linux/usb.h b/include/linux/usb.h
index 7d97179c3bf5..cf7b6bfc0838 100644
--- a/include/linux/usb.h
+++ b/include/linux/usb.h
@@ -1192,6 +1192,7 @@ struct usb_driver {
 	int (*unlocked_ioctl) (struct usb_interface *intf, unsigned int code,
 			void *buf);
 
+	int (*prepare) (struct usb_interface *intf);
 	int (*suspend) (struct usb_interface *intf, pm_message_t message);
 	int (*resume) (struct usb_interface *intf);
 	int (*reset_resume)(struct usb_interface *intf);
-- 
2.31.1

